uitools/widgets/Esri/ArcGISRuntime/Toolkit/AuthenticationView.cpp

```diff
@@ -0,0 +1,189 @@
+#include "Internal/UserCredentialView.h"
+#include "Internal/OAuth2View.h"
+#include "Internal/ClientCertificateView.h"
+#include "Internal/SslHandshakeView.h"
+
+namespace Esri
+{
+namespace ArcGISRuntime
+{
+namespace Toolkit
+{
+
+namespace 
+{
+  QWidget* createWidgetView(AuthenticationController* controller, AuthenticationView* view)
+  {
+    Q_ASSERT(controller);
+    auto challengeType = static_cast<AuthenticationChallengeType>(controller->currentChallengeType());
+
+    if (challengeType == AuthenticationChallengeType::UsernamePassword) {
+      // ArcGIS token, HTTP Basic/Digest, IWA
+      return new UserCredentialView(controller, view);
+    } 
+    else if (challengeType == AuthenticationChallengeType::OAuth)
+    {
+      // OAuth 2
+      return new OAuth2View(controller, view);
+    }
+    else if (challengeType == AuthenticationChallengeType::ClientCertificate)
+    {
+      // Client Certificate
+      return new ClientCertificateView(controller, view);
+    }
+    else if (challengeType == AuthenticationChallengeType::SslHandshake) {
+      // SSL Handshake - Self-signed certificate
+      return new SslHandshakeView(controller, view);
+    }
+    else
+    {
+      // Load nothing, challenge not understood.
+      return nullptr;
+    }
+  }
+}
+
+/*!
+  \class Esri::ArcGISRuntime::Toolkit::AuthenticationView
+  \inmodule EsriArcGISRuntimeToolkit
+  \ingroup ArcGISQtToolkitUiCppWidgetsViews
+  \since Esri.ArcGISRutime 100.12
+  \brief A view for handling authentication challenges and automatically 
+  launching the appropriate UI for each type of authentication.
+    
+  Declare an AuthenticationView. The AuthenticationView 
+  will then be connected to all authentication challenges, and will
+  automatically launch the appropriate view for the type of
+  challenge. Supported security formats include:
+
+  \list
+    \li ArcGIS Token (user credentials)
+    \li HTTP Basic (user credentials)
+    \li HTTP Digest (user credentials)
+    \li IWA (user credentials)
+    \li OAuth 2.0 (OAuth2)
+    \li SAML (OAuth2)
+    \li PKI (client certificate)
+    \li SSL Handshake Warnings (ssl)
+  \endlist
+  
+  \note OAuth 2.0 uses QWebEngineView. To use OAuth you must add \c{QT += webenginewidgets}
+        to \c{qmake}.
+        See \l{https://doc.qt.io/qt-5/qwebengineview.html}{QWebEngineView}.
+ */
+
+/*!
+  \brief Constructor.
+  \list
+    \li \a parent Parent widget.
+  \endlist
+
+  Internally the view will create and maintain its own \l AuthenticationController.
+ */
+AuthenticationView::AuthenticationView(QWidget* parent) :
+  AuthenticationView(nullptr, parent)
+{
+  setController(new AuthenticationController(this));
+}
+
+/*!
+  \brief Constructor.
+  \list
+    \li \a controller AuthenticationController which will drive the view.
+    \li \a parent Parent widget.
+  \endlist
+ */
+AuthenticationView::AuthenticationView(AuthenticationController* controller, QWidget* parent) :
+  QDialog(parent),
+  m_controller(controller),
```

Comment:
Is it make sense to create an `AuthenticationView` with null controller? For example if user writes:
```cpp
auto view = new AuthenticationView(nullptr, parent);
```
I suppose the `AuthenticationView` is not usable in this case.

Maybe this is better to avoid that?
```cpp
m_controller(controller ? controller : new AuthenticationController(this))
```

And in this case, you can replace all the test for the `m_controller` in the different views by assert.